ggml-sycl.cpp

```diff
@@ -15980,63 +16112,100 @@ static void ggml_sycl_mul_mat_id(const ggml_tensor *src0,
+                            k_copy_src1_to_contiguous(
+                                src1_original, src1_contiguous_get_ct1,
+                                dev_cur_src1_row_get_ct2,
+                                dev_row_mapping_get_ct3, ids_dev, i02,
+                                ids_nb_ct6, ids_nb_ct7, ne11, ne10, nb11, nb12,
+                                item_ct1, src1_row_acc_ct1);
+                        });
+                });
+            }
+
+            src0_row_extra.data_device[g_main_device] = src0_original + i02*nb02;
+
+            GGML_ASSERT(nb11 == sizeof(float)*ne10);
+            GGML_ASSERT(nb1 == sizeof(float)*ne0);
             src1_row.ne[1] = num_src1_rows;
-            dst_row.ne[1] = num_src1_rows;
 
             src1_row.nb[1] = nb11;
             src1_row.nb[2] = num_src1_rows*nb11;
             src1_row.nb[3] = num_src1_rows*nb11;
 
+            dst_row.ne[1] = num_src1_rows;
             dst_row.nb[1] = nb1;
             dst_row.nb[2] = num_src1_rows*nb1;
             dst_row.nb[3] = num_src1_rows*nb1;
 
             ggml_sycl_mul_mat(&src0_row, &src1_row, &dst_row);
 
-            num_src1_rows = 0;
-            for (int64_t i01 = 0; i01 < ids->ne[1]; i01++) {
-                const int32_t row_id_i = *(const int32_t *) (ids_host.data() + i01*ids->nb[1] + id*ids->nb[0]);
-
-                if (row_id_i != row_id) {
-                    continue;
-                }
-
-                GGML_ASSERT(row_id >= 0 && row_id < n_as);
-
-                SYCL_CHECK(CHECK_TRY_ERROR(stream->memcpy(
-                    dst_original + i01 * nb1,
-                    dst_contiguous.get() + num_src1_rows * nb1, nb1)));
-                num_src1_rows++;
+            {
+                sycl::range<3> block_dims(1, 1, std::min((unsigned int)ne0, 768u));
+                sycl::range<3> grid_dims(1, 1, num_src1_rows);
+                stream->submit([&](sycl::handler &cgh) {
+                    const char *__restrict dst_contiguous_get_ct1 =
+                        dst_contiguous.get();
+                    const mmid_row_mapping *__restrict dev_row_mapping_get_ct2 =
+                        dev_row_mapping.get();
+
+                    cgh.parallel_for(
+                        sycl::nd_range<3>(grid_dims * block_dims, block_dims),
+                        [=](sycl::nd_item<3> item_ct1) {
+                            k_copy_dst_from_contiguous(dst_original,
+                                                       dst_contiguous_get_ct1,
+                                                       dev_row_mapping_get_ct2,
+                                                       ne0, nb1, nb2, item_ct1);
+                        });
+                });
             }
         }
-    }
-
-    if (dst->backend == GGML_BACKEND_TYPE_CPU) {
-        SYCL_CHECK(CHECK_TRY_ERROR(stream->wait()));
+        if (dst->backend == GGML_BACKEND_TYPE_CPU) {
+            SYCL_CHECK(CHECK_TRY_ERROR(stream->wait()));
+        }
```

Comment:
seems cuda doesn't need this?

Reply:
Yes, rm them.

Reply:
> All uses of `ggml_tensor::backend` should be removed. The SYCL backend can assume that all tensors it receives are allocated in a SYCL buffer.

This PR is following current design.
There is another work (ongoing) to refactor the code as your comment: use context to save the backend.